server/handlers/authentication_test.go

```diff
@@ -0,0 +1,221 @@
+* 3) request.Cookie("token") != nil
+* 4) etc...
+ */
+
+func Test_GetLoginHandler_NoToken_OK(t *testing.T) {
+	// Prepare the test
+	var (
+		proxyUrl                  = "http://localhost:5001"
+		expectedLoginHTMLPath     = "assets-v1/templates/src/pages/oauth_portal/login.html"
+		expectedHTMLParsingParams = map[string]interface{}{
+			"HasNext":  true,
+			"Next":     "/",
+			"ProxyURL": proxyUrl,
+		}
+	)
+
+	ctrl := gomock.NewController(t)
+
+	serviceMock := mocks.NewMockServiceInterface(ctrl)
+	htmlParserMock := func(w http.ResponseWriter, htmlFile string, params map[string]interface{}) {
+		assert.Equal(t, expectedLoginHTMLPath, htmlFile)
+		assert.Equal(t, expectedHTMLParsingParams, params)
+	}
+
+	handler := NewAuthenticationHandler(serviceMock, htmlParserMock)
+
+	os.Setenv("PROXY_URL", proxyUrl)
+
+	// Execute the test
+	req := httptest.NewRequest("GET", "/login", nil)
+	responseRecorder := httptest.NewRecorder()
+	handler.Login(responseRecorder, req)
+
+	// Verify the results
+	assert.Equal(t, http.StatusOK, responseRecorder.Code)
+}
+
+func Test_GetLoginHandler_TokenExists_OK(t *testing.T) {
+
+	// Prepare the test
+	var (
+		fakeJwtToken = "fakeJwtToken"
+	)
+
+	ctrl := gomock.NewController(t)
+
+	serviceMock := mocks.NewMockServiceInterface(ctrl)
+	serviceMock.EXPECT().GetUserByToken(fakeJwtToken).Return(&models.User{}, nil)
+
+	handler := NewAuthenticationHandler(serviceMock, nil)
+
+	// Execute the test
+	req := httptest.NewRequest("GET", "/login", nil)
+	req.AddCookie(&http.Cookie{Name: "token", Value: fakeJwtToken})
+	responseRecorder := httptest.NewRecorder()
+	handler.Login(responseRecorder, req)
+
+	// Verify the results
+	assert.Equal(t, http.StatusSeeOther, responseRecorder.Code)
+	assert.Equal(t, "/", responseRecorder.Header().Get("Location"))
+}
+
+func Test_PostLoginHandler_ValidCredentials_OK(t *testing.T) {
+	// Prepare the test
+	var (
+		username     = "username"
+		password     = "password"
+		nextUrl      = "/next"
+		fakeJwtToken = "fakeJwt"
+
+		loginResult = map[string]interface{}{
+			"token":    fakeJwtToken,
+			"username": username,
+		}
+	)
+
+	ctrl := gomock.NewController(t)
+
+	serviceMock := mocks.NewMockServiceInterface(ctrl)
+	serviceMock.EXPECT().LoginUser(username, password).Return(loginResult, nil)
+
+	handler := NewAuthenticationHandler(serviceMock, nil)
+
+	// Execute the test
+	params := url.Values{}
+	params.Add("username", username)
+	params.Add("password", password)
+
+	req := httptest.NewRequest("POST", "/login", bytes.NewBufferString(params.Encode()))
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.URL.RawQuery = "next=" + nextUrl
+
+	responseRecorder := httptest.NewRecorder()
+	handler.Login(responseRecorder, req)
+
+	// Verify the results
+	assert.Equal(t, http.StatusSeeOther, responseRecorder.Code)
+	assert.Equal(t, nextUrl, responseRecorder.Header().Get("Location"))
+
+	var isCookieStored bool
+	for _, cookie := range responseRecorder.Result().Cookies() {
```

Comment:
(no need to change, but...)
Why range over all the cookies here? You are calling:
`responseRecorder.Result().Cookies()` but why not just `responseRecorder.Result().Cookie("token")`

Reply:
I think that is not possible because if you see at the `Cookies()` func, it is actually an array of `[]*http.Cookie`, and in order to access each field inside of it, doing a for loop is necessary here, after which we can only access each of the `http.Cookie` fields, and check `if cookie.Name == "token"`
![image](https://github.com/globe-and-citizen/layer8/assets/76698805/6be219e7-0eae-4699-be6c-850686e272cf)

Reply:
Yes huzaifa was correct